modules/storages/app/common/storages/peripherals/one_drive_connection_validator.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal:true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    class OneDriveConnectionValidator
+      include Dry::Monads[:maybe]
+
+      using ServiceResultRefinements
+
+      def initialize(storage:)
+        @storage = storage
+      end
+
+      def validate
+        maybe_is_not_configured
+          .or { tenant_id_wrong }
+          .or { client_id_wrong }
+          .or { client_secret_wrong }
+          .or { drive_id_wrong }
+          .or { request_failed_with_unknown_error }
+          .or { drive_with_unexpected_content }
+          .value_or(ConnectionValidation.new(type: :healthy, timestamp: Time.current, description: nil))
```

Comment:
🟢 The monads are a really nice touch, but I wonder if we can follow a positive naming scheme- the negation based naming can sometimes be confusing (at least to me)
```suggestion
        maybe_storage_configured
          .or { validate_tenant_id }
          .or { validate_client_id }
          .or { validate_client_secret }
          .or { validate_drive_id }
          .or { request_failed_with_unknown_error }
          .or { drive_with_unexpected_content }
          .value_or(ConnectionValidation.new(type: :healthy, timestamp: Time.current, description: nil))
```

Reply:
Ah, well, that is interesting. Especially due to the `or` concatenation I naturally read negative namings. It seems more natural then the positive way.

See this imaginary monologue.

"So, I have this storage.... I suspect it to be invalid."
"Maybe the storage is not fully configured?"
"Well, apparently it is... so, is then tenant id wrong?"
"Hmm, not... what about the client id?"
"Or the client secret?"
"Anything of that? No? Then apparently you are valid."

Get what I am pointing at? :D validations are "positive" when they are "negative", hence finding a problem means, your validation was justified.

Reply:
> Get what I am pointing at? :D validations are "positive" when they are "negative", hence finding a problem means, your validation was justified.

I see wym for sure, and perhaps ActiveModel::Validations are a great conventional example; they only push to `errors` in the even "negative" event that there's a validation failure, otherwise they just return ?true.

Hence in your imaginany monologue 😄 mine would read like so:

"So I have this storage... I need to validate that everything is configured properly"
"Firstly, validated it is fully configured"
"Then, validate the tenant id is correct"
"Then, validate the client id is correct"
"Then, validate the client secret is correct"
"When there are invalid cases, collect them and display a unified response (similar to active_model_errors) to the user"
"When there are no errors, return- mark the storage connection as valid"

Reply:
> but AFAICT- we fail fast on the first error result? 

Here's the trick, except for the missing/empty values, it is impossible to validate if the previous validation has failed. If the tenant id is wrong, we can't validate the Client ID/Secret, if the Client ID/Secret is wrong we can't validate the Drive ID.

> In which case, I wonder if just mixin in [ActiveModel::Validations](https://api.rubyonrails.org/classes/ActiveModel/Validations.html) and writing these as validate statements would be more fitting albeit less "Monad-ish" 😄

There are always more than one way to do things. A `dry/validation` contract, `ActiveModel::Validations` or a chain of assertions all could do the same thing.

1. ActiveModel would be tied directly to an object (with certain assumptions). 
2. `Dry::Validation::Contract` would create an object just for validation of a hash/json that could be passed down after validation.
3. A chain of assertions (in this case using the Maybe/Option monad) makes explicit the validation process.

Reply:
> Here's the trick, except for the missing/empty values, it is impossible to validate if the previous validation has failed. If the tenant id is wrong, we can't validate the Client ID/Secret, if the Client ID/Secret is wrong we can't validate the Drive ID.

Yes, and here we should design it, as the domain user story is telling us. And here, it was set, we do not show all errors. We do show the first error, that can happen. Hence, the implementation follows that story, and fails with the first error, not checking for the other validation rules.